File: src/m/miscellaneous/normfit_issm.m
%
%  wrapper for normfit to avoid using the matlab statistics toolbox.
%
function [muhat,sigmahat,muci,sigmaci]=normfit_issm(x,alpha)

	if ~exist('alpha','var')
		alpha=0.05;
	end

%  check for any NaN in any columns

	if ~any(any((isnan(x))))

%  explicitly calculate the moments

		muhat   =mean(x);
		sigmahat=std(x);

		if (nargout>2)
			prob=1.-alpha/2.;

			if (size(x,1) == 1)
				% operate like matlab normfit, mean, std, etc.
				n=length(x);
			else
				n=size(x,1);
			end

			muci    =zeros(2,length(muhat   ));
			sigmaci =zeros(2,length(sigmahat));

			try
				muci(1,:)   =muhat-tinv(prob,n-1)*sigmahat/sqrt(n);
				muci(2,:)   =muhat+tinv(prob,n-1)*sigmahat/sqrt(n);
				sigmaci(1,:)=sigmahat*sqrt((n-1)/chi2inv(prob   ,n-1));
				sigmaci(2,:)=sigmahat*sqrt((n-1)/chi2inv(1.-prob,n-1));
			catch me
				muci(1,:)   =muhat;
				muci(2,:)   =muhat;
				sigmaci(1,:)=sigmahat;
				sigmaci(2,:)=sigmahat;
			end
		end

	else

%  must loop over columns, since number of elements could be different

		muhat   =zeros(1,size(x,2));
		sigmahat=zeros(1,size(x,2));
		muci    =zeros(2,size(x,2));
		sigmaci =zeros(2,size(x,2));

%  remove any NaN and recursively call column

		for j=1:size(x,2)
			[muhat(j),sigmahat(j),muci(:,j),sigmaci(:,j)]=normfit_issm(x(~isnan(x(:,j)),j),alpha);
		end
	end
end

File: src/m/miscellaneous/normfit_issm.py
import numpy as np
from scipy.stats import chi2, t
#  wrapper for normfit to avoid using the matlab statistics toolbox.


def normfit_issm(x, alpha=None):
    if alpha is None:
        alpha = 0.05

    #  check for any NaN in any columns
    if not np.isnan(x).any():

        #  explicitly calculate the moments
        muhat = np.mean(x, 0)
    # numpy defaults to 0 delta degrees of freedom; matlab uses 1
        sigmahat = np.std(x, 0, ddof=1)

    # no way to ask this in python, assume 4 outputs
    #if (nargout > 2):
        prob = 1. - alpha / 2.

        if (np.size(x, 0) == 1):
            # operate like matlab normfit, mean, std, etc.
            n = np.size(x)
        else:
            n = np.size(x, 0)

        muci = np.zeros((2, np.size(muhat)))
        sigmaci = np.zeros((2, np.size(sigmahat)))

        try:
            muci[0, :] = muhat - t.ppf(prob, n - 1) * sigmahat / np.sqrt(n)
            muci[1, :] = muhat + t.ppf(prob, n - 1) * sigmahat / np.sqrt(n)
            sigmaci[0, :] = sigmahat * np.sqrt((n - 1) / chi2.ppf(prob, n - 1))
            sigmaci[1, :] = sigmahat * np.sqrt((n - 1) / chi2.ppf(1. - prob, n - 1))
        except:
            muci[0, :] = muhat
            muci[1, :] = muhat
            sigmaci[0, :] = sigmahat
            sigmaci[1, :] = sigmahat
    else:
        #  must loop over columns, since number of elements could be different
        muhat = np.zeros((1, np.size(x, 1)))
        sigmahat = np.zeros((1, np.size(x, 1)))
        muci = np.zeros((2, np.size(x, 1)))
        sigmaci = np.zeros((2, np.size(x, 1)))

    #  remove any NaN and recursively call column
        for j in range(np.shape(x, 1)):
            [muhat[j], sigmahat[j], muci[:, j], sigmaci[:, j]] = normfit_issm(x[not np.isnan(x[:, j]), j], alpha)

    return [muhat, sigmahat, muci, sigmaci]
